server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContent.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+/**
+ * An extension of {@link ToXContent} that can be serialized in chunks by creating a {@link ChunkedXContentSerialization}.
+ * This is used by the REST layer to implement flow control that does not rely on blocking the serializing thread when writing the
+ * serialized bytes to a non-blocking channel.
+ */
+public interface ChunkedToXContent extends ToXContent {
+
+    /**
+     * Create a serialization session for the implementation instance that will flush to the given {@code builder} and use the given
+     * {@code params} the same way {@link ToXContent#toXContent(XContentBuilder, Params)} would.
+     * @param builder builder to flush to
+     * @param params params for serialization
+     * @return serialization session
+     */
+    ChunkedXContentSerialization toXContentChunked(XContentBuilder builder, ToXContent.Params params);
+
+    @Override
+    default XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        ChunkedXContentSerialization serialization = toXContentChunked(builder, params);
```

Comment:
Can we add a minimal test of this default implementation?

Reply:
Seems kind of redundant doesn't it? We currently test this for the one implementation we have in `org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsResponseTests#testToChunkedXContent`. What does it add if I add a specific test with some (I presume) made up chunked x-content implementation?